launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java

```diff
@@ -92,18 +92,11 @@
   List<String> buildJavaCommand(String extraClassPath) throws IOException {
     List<String> cmd = new ArrayList<>();
 
-    String[] candidateJavaHomes = new String[] {
-      javaHome,
+    String firstJavaHome = firstNonEmpty(javaHome,
       childEnv.get("JAVA_HOME"),
       System.getenv("JAVA_HOME"),
-      System.getProperty("java.home")
-    };
-    for (String javaHome : candidateJavaHomes) {
-      if (javaHome != null) {
-        cmd.add(join(File.separator, javaHome, "bin", "java"));
-        break;
-      }
-    }
+      System.getProperty("java.home"));
+    cmd.add(join(File.separator, firstJavaHome, "bin", "java"));
```

Comment:
Technically saying, this is not 100% same. We need to deal with the condition `firstJavaHome == null`. I wouldn't say it's a normal case, but just to be safe.

Reply:
Yes, i have considered of `firstJavaHome == null`, but origin code logic not deal this situation here, it will throw exception at higher caller, so not add `if (fisrstJavaHome == null)` to keep the origin code logic.

Reply:
Actually it will be dealed in `join`.....Anyway, have added this, thanks for review.